NODE_PATH points into a non-existent directory when pnpm runs from a workspace package

### Last pnpm version that worked

10.7.1

### pnpm version

11.21.0

### Code to reproduce the issue

https://github.com/Merciful12/pnpm-regression

`preferSymlinkedExecutables: true` makes pnpm export `NODE_PATH` so scripts can resolve packages from the hoisted virtual store. The reproduction relies on `ansi-styles`, a transitive dependency of `chalk`, which exists only there:

```
node_modules/chalk
node_modules/.pnpm/node_modules/ansi-styles   <- only copy, reachable via NODE_PATH
```

```sh
pnpm install

# from the workspace root
pnpm -F app node-path
pnpm -F app check          # exit 0

# from the package directory
cd packages/app
pnpm node-path
pnpm check                 # exit 1
```

Both config files in the repo are load-bearing: 10.x reads `prefer-symlinked-executables` only from `.npmrc`, 11.x only from `pnpm-workspace.yaml`.

### Expected behavior

`NODE_PATH` points at the hoisted store of the workspace root regardless of the directory pnpm was invoked from, so both invocations behave identically:

```
<repro>/node_modules/.pnpm/node_modules
```

This is what pnpm 10.7.1 and earlier do — `pnpm check` exits 0 from both locations.

### Actual behavior

When pnpm is invoked from inside a workspace package, `NODE_PATH` is built from that package's directory instead of the workspace root:

```
<repro>/packages/app/node_modules/.pnpm/node_modules
```

That directory does not exist — the virtual store lives only at the workspace root. Every dependency reachable only through the hoisted store then fails to resolve:

```
Error: Cannot find module 'ansi-styles'
```

| pnpm | invoked from | NODE_PATH | `pnpm check` |
|---|---|---|---|
| 10.7.1 | workspace root | `<repro>/node_modules/.pnpm/node_modules` | exit 0 |
| 10.7.1 | `packages/app` | `<repro>/node_modules/.pnpm/node_modules` | exit 0 |
| 10.8.0 | workspace root | `<repro>/node_modules/.pnpm/node_modules` | exit 0 |
| 10.8.0 | `packages/app` | `<repro>/packages/app/node_modules/.pnpm/node_modules` | **exit 1** |
| 11.21.0 | `packages/app` | `<repro>/packages/app/node_modules/.pnpm/node_modules` | **exit 1** |

### Additional information

Bisected over all 77 published `10.x` releases: last working `10.7.1`, first broken `10.8.0`. Still present in `11.21.0`.

`pnpm root -w` from the same directory correctly prints `<repro>/node_modules`, and filters resolve normally — pnpm does know the workspace root, it just isn't reflected in the config used to build `extraEnv`. Setting `sharedWorkspaceLockfile: true` explicitly changes nothing, so `workspaceDir` rather than that flag is what is missing.

The expression that builds `NODE_PATH` is byte-for-byte identical in `10.7.1` and `10.8.0` (`dist/pnpm.cjs`), and unchanged in `11.20.0` (`dist/pnpm.mjs`):

```js
if (pnpmConfig.preferSymlinkedExecutables && !isWindows()) {
  const cwd = pnpmConfig.lockfileDir ?? pnpmConfig.dir;
  const virtualStoreDir = pnpmConfig.virtualStoreDir
    ? pnpmConfig.virtualStoreDir
    : pnpmConfig.modulesDir ? path.join(pnpmConfig.modulesDir, '.pnpm') : 'node_modules/.pnpm';
  pnpmConfig.extraEnv['NODE_PATH'] = pathAbsolute(path.join(virtualStoreDir, 'node_modules'), cwd);
}
```

So the regression looks to be in how `workspaceDir` / `lockfileDir` are populated when the cwd is a workspace package, not in the expression itself. 10.8.0 shipped the `updateConfig` pnpmfile hook (#9325) and a config-writer refactor (#9306) — I have not confirmed which commit.

Any package script that shells out to Node and relies on the hoisted store now depends on where it was started, including scripts that invoke a nested `pnpm`, which runs with the package directory as its cwd.

### Node.js version

24.19.0

### Operating System

macOS

--- a/.changeset/node-path-uses-the-workspace-virtual-store.md
+++ b/.changeset/node-path-uses-the-workspace-virtual-store.md
@@ -3,4 +3,4 @@
 "pnpm": patch
 ---
 
-With `preferSymlinkedExecutables`, `NODE_PATH` again points at the virtual store of the workspace root when pnpm is run from inside a workspace package. It was built from the package directory instead, so scripts could not resolve anything that only lives in the hoisted store [#13912](https://github.com/pnpm/pnpm/issues/13912).
+With `preferSymlinkedExecutables`, `NODE_PATH` again points at the virtual store of the workspace root when pnpm is run from inside a workspace package, so scripts can resolve dependencies that live only in the hoisted store [#13912](https://github.com/pnpm/pnpm/issues/13912).

--- a/.changeset/node-path-uses-the-workspace-virtual-store.md
+++ b/.changeset/node-path-uses-the-workspace-virtual-store.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/config.reader": patch
+"pnpm": patch
+---
+
+With `preferSymlinkedExecutables`, `NODE_PATH` again points at the virtual store of the workspace root when pnpm is run from inside a workspace package. It was built from the package directory instead, so scripts could not resolve anything that only lives in the hoisted store [#13912](https://github.com/pnpm/pnpm/issues/13912).

--- a/pnpm11/config/reader/src/index.ts
+++ b/pnpm11/config/reader/src/index.ts
@@ -684,6 +684,10 @@ export async function getConfig (opts: {
     pnpmConfig.filterProd = (pnpmConfig.filterProd as string).split(' ')
   }
 
+  if (pnpmConfig.sharedWorkspaceLockfile && !pnpmConfig.lockfileDir && pnpmConfig.workspaceDir) {
+    pnpmConfig.lockfileDir = pnpmConfig.workspaceDir
+  }
+
   if (pnpmConfig.workspaceDir) {
     pnpmConfig.extraBinPaths = [path.join(pnpmConfig.workspaceDir, 'node_modules', '.bin')]
   } else {
@@ -763,10 +767,6 @@ export async function getConfig (opts: {
   }
   pnpmConfig.sideEffectsCacheRead = pnpmConfig.sideEffectsCache ?? pnpmConfig.sideEffectsCacheReadonly
   pnpmConfig.sideEffectsCacheWrite = pnpmConfig.sideEffectsCache
-
-  if (pnpmConfig.sharedWorkspaceLockfile && !pnpmConfig.lockfileDir && pnpmConfig.workspaceDir) {
-    pnpmConfig.lockfileDir = pnpmConfig.workspaceDir
-  }
 
   pnpmConfig.workspaceConcurrency = getWorkspaceConcurrency(pnpmConfig.workspaceConcurrency)
 

--- a/pnpm11/config/reader/test/index.ts
+++ b/pnpm11/config/reader/test/index.ts
@@ -10,6 +10,7 @@ import { esmNodePathLoaderImportFlag } from '@pnpm/exec.esm-node-path-loader'
 import { prepare, prepareEmpty } from '@pnpm/prepare'
 import { fixtures } from '@pnpm/test-fixtures'
 import { isCI } from 'ci-info'
+import isWindows from 'is-windows'
 import PATH from 'path-name'
 import { symlinkDir } from 'symlink-dir'
 import { writeYamlFileSync } from 'write-yaml-file'
@@ -41,6 +42,7 @@ const env = {
   [PATH]: path.join(import.meta.dirname, 'bin'),
 }
 const f = fixtures(import.meta.dirname)
+const testOnPosix = isWindows() ? test.skip : test
 
 test('getConfig()', async () => {
   const { config } = await getConfig({
@@ -3759,6 +3761,28 @@ test('preferSymlinkedExecutables should be true when nodeLinker is hoisted', asy
     },
   })
   expect(config.preferSymlinkedExecutables).toBeTruthy()
+})
+
+testOnPosix('NODE_PATH points to the virtual store of the workspace root when pnpm runs from a workspace package', async () => {
+  prepareEmpty()
+
+  const workspaceDir = process.cwd()
+  const pkgDir = path.join(workspaceDir, 'packages/app')
+  fs.mkdirSync(pkgDir, { recursive: true })
+
+  const { config } = await getConfig({
+    cliOptions: {
+      dir: pkgDir,
+      'prefer-symlinked-executables': true,
+    },
+    packageManager: {
+      name: 'pnpm',
+      version: '1.0.0',
+    },
+    workspaceDir,
+  })
+
+  expect(config.extraEnv['NODE_PATH']).toBe(path.join(workspaceDir, 'node_modules/.pnpm/node_modules'))
 })
 
 test('return a warning when the .npmrc has an env variable that does not exist', async () => {

